### PR TITLE
Propagate noop_with_empty_axes in reduce operators.

### PR DESCRIPTION
**Pull request by AtanasDimitrovQC**
### Description
Since opset 18 `reduce` operators have a `noop_with_empty_axes` flag. This flag was previously not being propagated in most of the `reduce` operators(#16586). 
This pull request: 
1. Fixes flag-propagation for `reduce` operators.
2. Adds tests for preventing future deviations from the standart.

### CODE REVIEW

**Comment by AtanasDimitrovQC** [comment]
@microsoft-github-policy-service agree company="QuantCo"

**Comment by AtanasDimitrovQC** [comment]
> @AtanasDimitrovQC please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="Microsoft"

**Comment by AtanasDimitrovQC** [comment]
> @AtanasDimitrovQC please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="QuantCo"

**Review by xadupre — APPROVED** [review]

File: onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc [hunk 1]
```diff
@@ -2425,7 +2425,7 @@ TEST(ReductionOpTest, ReduceSum_noop_axes_input_initializer) {
   test.Run(OpTester::ExpectResult::kExpectSuccess, "", {kTensorrtExecutionProvider, kOpenVINOExecutionProvider});
 }
 
-TEST(ReductionOpTest, ReduceSum_empty_axes_input_initializer) {
+TEST(ReductionOpTest, ReduceSum_13_empty_axes_input_initializer) {
```
[the remark this answers is not in the document]
Comment (AtanasDimitrovQC) [reply onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc]: Fixed the inconsistency. 
This said as far as I can see there is no general rule about naming the tests in the test-suite(e.g. there are both snake-cased ones as well as camel-cased ones).

**Review by xadupre — APPROVED** [review]

File: onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (edgchen1) [reply onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc]: just FYI, regarding naming, Google Test does have a rule.
https://google.github.io/googletest/faq.html#why-should-test-suite-names-and-test-names-not-contain-underscore

unfortunately, we are not very good about following it

**Comment by AtanasDimitrovQC** [comment]
Is there anything else I can do to get it merged?

**Comment by adityagoel4512** [comment]
@baijumeswani would it be possible to merge this PR please? It would be very useful to have before the next onnxruntime release.

**Comment by baijumeswani** [comment]
[AtanasDimitrovQC](https://github.com/AtanasDimitrovQC), [adityagoel4512](https://github.com/adityagoel4512) the CI pipelines seem to be failing. Could you please address the issue?

tagging the DRIs for this week for a review/follow-up: @zhijxu-MS @fs-eire

**Comment by AtanasDimitrovQC** [comment]
@baijumeswani 
I removed testing for all the providers that seem to implement the `noop_with_empty_axes` flag incorrectly. This seems to be a general practice of the whole reduction test suite.

**Review by baijumeswani — APPROVED** [review]

**Comment by baijumeswani** [comment]
Thanks [AtanasDimitrovQC](https://github.com/AtanasDimitrovQC) for your contribution.

**Comment by snnn** [comment]
This PR seems causing Android CI pipeline to fail. And nobody has run this pipeline before this PR is merged.

File: onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc [hunk 2]
```diff
@@ -3373,6 +3373,241 @@ TEST(ReductionOpTest, ReduceSum_ReduceDimWithZero3) {
...
+                        3.0f, 4.0f});
+  test.AddInput<int64_t>("axes", {0}, {}, true);
+  test.AddOutput<float>("reduced", {1, 2, 2}, {1.0f, 2.0f, 3.0f, 4.0f});
+  test.Run(
+      OpTester::ExpectResult::kExpectSuccess,
+      "",
+      {kTensorrtExecutionProvider,
+       kOpenVINOExecutionProvider,
+       kDnnlExecutionProvider,
+       kDmlExecutionProvider});
```
Comment (edgchen1) [inline onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc:3392]: @jywu-msft @fdwr
FYI about some newly added tests which have been excluded for these EPs.
Taking a guess at the point of contact for these EPs. If it's someone else, please let them know.

Comment (fdwr) [reply onnxruntime/test/providers/cpu/reduction/reduction_ops_test.cc:3392]: @linnealovespie I think you added `noop_axes` recently to the DML EP reduction operators, right?

**Pull request merged**